src/compiler/program.ts

```diff
@@ -904,7 +904,13 @@ namespace ts {
             let declarationDiagnostics: Diagnostic[] = [];
 
             if (options.noEmit) {
-                return { diagnostics: declarationDiagnostics, sourceMaps: undefined, emittedFiles: undefined, emitSkipped: true };
+                // Allow only emitting of declarations
+                if (options.declaration) {

```

Comment:
But are there cases where you want to ensure that declaration emit can succeed? For example, just something where you want to ensure that `declaration` won't prompt an error because of non-exported declarations?